File: app/src/main/java/com/qianlei/jiaowu/ui/fragment/setting/SettingFragment.java
package com.qianlei.jiaowu.ui.fragment.setting;

import android.os.Bundle;
import android.widget.Toast;

import androidx.preference.EditTextPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.qianlei.jiaowu.MainApplication;
import com.qianlei.jiaowu.R;
import com.qianlei.jiaowu.db.MyDataBase;

import java.io.File;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;


/**
 * @author qianlei
 */
public class SettingFragment extends PreferenceFragmentCompat {

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.fragement_setting, rootKey);
        EditTextPreference editTextPreference = findPreference("start_day");
        if (editTextPreference != null) {
            editTextPreference.setSummary(editTextPreference.getText());
            editTextPreference.setOnPreferenceChangeListener((preference, newValue) -> {
                if (newValue instanceof String) {
                    String s = (String) newValue;
                    DateFormat dateFormat = SimpleDateFormat.getDateInstance(SimpleDateFormat.MEDIUM, Locale.CHINA);
                    try {
                        System.out.println(dateFormat.format(new Date()));
                        dateFormat.parse(s);
                        editTextPreference.setSummary(s);
                        return true;
                    } catch (ParseException e) {
                        Toast.makeText(getContext(), "日期填写错误,格式为yyyy年MM月dd日", Toast.LENGTH_SHORT).show();
                        return false;
                    }
                }
                return true;
            });
        }
        Preference preference = findPreference("clear_log");
        if (preference != null) {
            preference.setOnPreferenceClickListener(preference1 -> {
                Toast.makeText(MainApplication.getInstance(), "清除缓存成功", Toast.LENGTH_SHORT).show();
                new Thread(() -> {
                    //删除缓存文件
                    File file = MainApplication.getInstance().getExternalCacheDir();
                    if (file != null && file.exists() && file.isDirectory()) {
                        File[] files = file.listFiles();
                        if (files != null) {
                            for (File f : files) {
                                //noinspection ResultOfMethodCallIgnored
                                f.delete();
                            }
                        }
                    }
                    //从数据库中删除所有的数据
                    MyDataBase dataBase = MyDataBase.getDatabase(MainApplication.getInstance());
                    dataBase.getExamDao().deleteAll();
                    dataBase.getScoreDao().deleteAll();
                    dataBase.getSubjectDao().deleteAll();
                }).start();
                return true;
            });
        }


    }
}

File: app/src/main/java/com/qianlei/jiaowu/ui/fragment/setting/SettingFragment.kt
package com.qianlei.jiaowu.ui.fragment.setting

import android.app.DatePickerDialog
import android.os.Build
import android.os.Bundle
import android.widget.DatePicker
import android.widget.Toast
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.afollestad.date.dayOfMonth
import com.afollestad.date.month
import com.afollestad.date.year
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.datetime.datePicker
import com.qianlei.jiaowu.R
import com.qianlei.jiaowu.db.MyDataBase
import com.qianlei.jiaowu.repository.HitokotoRepository
import com.qianlei.jiaowu.repository.SettingRepository
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.File
import java.util.*


/**
 * @author qianlei
 */
class SettingFragment : PreferenceFragmentCompat() {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.fragement_setting, rootKey)
        //选择开学日期
        val startDayPreference = findPreference<Preference>(getString(R.string.start_day))
        if (startDayPreference != null) {
            setStartDaySummary()
            startDayPreference.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                return@OnPreferenceClickListener changeStartDay()
            }
        }
        //清楚缓存
        val clearCachePreference = findPreference<Preference>(getString(R.string.clear_log))
        if (clearCachePreference != null) {
            clearCachePreference.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                return@OnPreferenceClickListener deleteCache()
            }
        }
        val hitokotoPreference = findPreference<Preference>(getString(R.string.hitokoto))
        if (hitokotoPreference != null) {
            hitokotoPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
                return@OnPreferenceChangeListener changeHitokoto(newValue)
            }
        }
    }

    private fun changeHitokoto(newValue: Any?): Boolean {
        if (newValue is Boolean) {
            if (newValue) {
                HitokotoRepository.getPoem()
            } else {
                HitokotoRepository.hitokotoLiveData.value = ""
            }
        }
        return true
    }

    /**
     * 设置开始时间的简介部分
     */
    private fun setStartDaySummary() {
        val c = context ?: return
        val startDayPreference = findPreference<Preference>(getString(R.string.start_day)) ?: return
        startDayPreference.summary = SettingRepository.getStartDayFormat(c)
    }

    /**
     * 修改开始时间
     */
    private fun changeStartDay(): Boolean {
        val c = context ?: return false
        val date = SettingRepository.getStartDay(c)
        val calendar = Calendar.getInstance()
        calendar.time = date

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val datePickerDialog = DatePickerDialog(c)
            datePickerDialog.updateDate(calendar.year, calendar.month, calendar.dayOfMonth)
            datePickerDialog.setOnDateSetListener { _: DatePicker, year: Int, month: Int, day: Int ->
                SettingRepository.setStartDay(c, year, month, day)
                setStartDaySummary()
            }
            datePickerDialog.show()
        } else {
            MaterialDialog(c).show {
                datePicker(currentDate = calendar) { _, date ->
                    SettingRepository.setStartDay(c, date.year, date.month, date.dayOfMonth)
                    setStartDaySummary()
                }
            }
        }

        return true
    }

    /**
     * 删除缓存
     */
    private fun deleteCache(): Boolean {
        val c = context ?: return false
        Toast.makeText(c, "清除缓存成功", Toast.LENGTH_SHORT).show()
        GlobalScope.launch {
            //删除错误记录文件
            val file: File? = c.externalCacheDir
            if (file != null && file.exists() && file.isDirectory) {
                val files = file.listFiles()
                if (files != null) {
                    for (f in files) {
                        f.delete()
                    }
                }
            }
            //从数据库中删除所有的数据
            val dataBase: MyDataBase = MyDataBase.getDatabase(context)
            dataBase.examDao().deleteAll()
            dataBase.scoreDao().deleteAll()
            dataBase.subjectDao().deleteAll()
        }
        return true
    }
}